### ISSUE TITLE
Fails collecting tests if test module level obj has __getattr__()

### ISSUE THREAD

**Issue by Suor** [pytest-dev/pytest#1035]
The issue is in [this line](https://github.com/pytest-dev/pytest/blob/dfd83b59f86ae49abe786deed5c97df00294bea1/_pytest/python.py#L362). Object returns something arbitrary and then it's required to be int lineno.

Breaks [this](https://travis-ci.org/Suor/funcy/jobs/81358215) for example.

**Comment by alfredodeza** [comment]
This only happens for me on Python 2.6. Python 2.7 is unaffected.

**Comment by nicoddemus** [comment]
Thanks for the report! :smile:

**Comment by RonnyPfannschmidt** [comment]
Since its only a proble!m on python 2.6 I'm considering leaving it be

**Comment by Suor** [comment]
So pytest no longer supports Python 2.6?

**Comment by RonnyPfannschmidt** [comment]
its still supported, but since python 2.6 is eol'd since years we should think about how much code and maintenance burdens we invest in it as volunteer work, in particular with libs that are somehow broken only on it

@hpk42 whats  your oppinion on that matter

**Comment by RonnyPfannschmidt** [comment]
just discussed with @hpk42 on irc

result: since we officially support it we should fix it

**Comment by jantman** [comment]
FWIW, I can confirm what @alfredodeza said - https://travis-ci.org/jantman/awslimitchecker/builds/82210772 tests against 26, 27, 32, 33, 34, pypy and pypy3; only 2.6 breaks.

**Comment by RonnyPfannschmidt** [comment]
at unit-test level all python packages seem to be required

**Comment by RonnyPfannschmidt** [comment]
i finished a first attempt to solving it, resulting in the python interpreter oom-ing my machine,

**Comment by The-Compiler** [comment]
It seems this is caused by 3497aa0766ac6f2f2cd66cc483c86f2a67a738aa / #921 

/cc @tomviner

**Comment by qwcode** [comment]
to be clear, this started with 2.8.0, correct?

**Comment by The-Compiler** [comment]
Yes, it seems to work fine with 2.7.3 which is before that commit.

**Comment by tomviner** [comment]
The difference in Python versions seems to be with `inspect.isclass`:

```python
$ python2.7
Python 2.7.9 (default, Apr  2 2015, 15:33:21)
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from inspect import isclass
>>> from whatever import _
>>> isclass(_)
False
```

which is the correct answer btw - `_` is `<whatever.Whatever object at 0x7ff296b57210>`, which is an instance not a class. But in Python 2.6:


```python
$ python2.6
Python 2.6.9 (default, Apr 16 2015, 18:31:03)
[GCC 4.9.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from inspect import isclass
>>> from whatever import _
>>> isclass(_)
True
```

[My commit](https://github.com/pytest-dev/pytest/commit/3497aa0766ac6f2f2cd66cc483c86f2a67a738aa), mentioned above, amends the `collector.istestclass` line below, but as you can see it's the `isclass` line above it that's, in the first instance, behaving erroneously:

```python
def pytest_pycollect_makeitem(collector, name, obj):
    ...
    if isclass(obj):
        if collector.istestclass(obj, name):
            Class = collector._getcustomclass("Class")
            outcome.force_result(Class(name, parent=collector))
```

See [pytest_pycollect_makeitem](https://github.com/pytest-dev/pytest/blob/cfd259ae6f839f32799254a4e3c902ec6bccb9ac/_pytest/python.py#L299).

So I propose the priority fix, is for us to provide a fixed `isclass` function when Python version == 2.6. I'd be happy to write a patch for that.

There's a second, and more questionable issue raised here. In my patch, I added support to respect nose's `Test.__test__` attribute to mark a class as a test despite not matching any naming rule. But is a class that responds `True` to `getattr(cls,  anything)` really signalling anything about tests?

The good news is that if we fix the first issue, we're not forced to answer the second.

**Comment by nicoddemus** [comment]
Thanks for the investigation work @tomviner! It seems your solution of using a fixed `isclass` function for Python 2.6 would work just fine and would not break anything for Python versions > 2.6, so I'm :+1: on it.

**Comment by The-Compiler** [comment]
The probably related CPython commit: https://hg.python.org/cpython/rev/35bf8f7a8edc?revcount=1920

**Comment by tomviner** [comment]
Comments on this patch please https://github.com/pytest-dev/pytest/pull/1121

**Comment by nicoddemus** [comment]
@Suor, could you please try #1121 in your test suite to ensure it works and does not break anything else by accident?

**Comment by The-Compiler** [comment]
@alfredodeza @jantman @renstrom @daikiri @jmoldow @trhoden any chance some of you could test that PR as well? :wink:

**Comment by jantman** [comment]
@The-Compiler This looks good to me - https://travis-ci.org/jantman/awslimitchecker/builds/84219885 and https://github.com/jantman/awslimitchecker/tree/pytest-pr-1121

**Comment by dairiki** [comment]
PR  #1121 works for me.  Thank you all!

**Comment by nicoddemus** [comment]
Fixed in #1121
